### PR TITLE
Add FW upgrade write-up

### PR DESCRIPTION
**Pull request by stulle123**
Hi @vegantransistor,

To keep you busy over the weekend, here's another write-up for you to review 😃 

Cheers,

stulle123

### CODE REVIEW

File: fw_upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+## Implementation
+
+The firmware upgrade is mainly implemented in the `/service_manager/services/utils/upgrade_write.pyc` file. This is what's going on:
+
+1. `upgrade_write()` --> decrypts the upgrade file, writes a block, gets the `uimage` and `uboot` signatures from the firmware file *or* calls `upgrade_write_init()`
+2. `upgrade_write_init()` --> header and CRC checks, writes a new SHA512 hash to flash memory
+3. `_write()` --> writes data to memory flash
+4. `upgrade_validate()` --> validates the file and checks the signature by calling `upgrade_validate_signature()`
+
+If the version string in the firmware header matches a certain value, the signature verification can be bypassed:
```
Comment (vegantransistor) [inline fw_upgrade/README.md:16]: maybe better write "signature is ignored" than "signature verification can be bypassed"

File: fw_upgrade/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+It's generated in the following way:
+
+1. The shared lib `_aes.cpython-38m.so` uses [EVP_BytesToKey](https://www.openssl.org/docs/manmaster/man3/EVP_BytesToKey.html) to generate the AES key and IV from a passphrase
+2. The passphrase can be found in ther file `upgrade_write.pyc` (it's `1.57079628`)
+3. Next, the passphrase is used to generate a "pre-key". This can be done with the openssl tool:
+```bash
+$ openssl enc -nosalt -p -in foo -out bar -e -aes256 -k "1.57079628" -md sha1
+```
+4. Next, that "pre-key" is used to decrypt a base64 string. The resulting "plaintext" makes a new passphrase: `LL0QuKHDBrOKeGHUSvUTVpwbHB2WzNcP`
```
Comment (vegantransistor) [inline fw_upgrade/README.md:37]: I would not publish the passphrase but let it as an exercise for the reader ;-)

File: fw_upgrade/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+1. Fire up an emulated Cradlepoint router as described [here](../router_emulation/README.md)
+2. Next, just run the following script with the `cppython` interpreter:
+```python
+from _aes import decryptobj, decrypt
+from math import atan
+import base64
+
+_KEY = str(atan(22041962))[0:10]
+pre_passphrase = decryptobj(_KEY)
+new_passphrase = pre_passphrase.decrypt(base64.b64decode(b'aqvFaN8nJ7JmKTH+mIZmo4qwhAZ8+/YrKzh//dLHR8+1gauSwKvuwCnVSp98n92x'))
```
Comment (vegantransistor) [inline fw_upgrade/README.md:54]: see comment above

**Review by vegantransistor — CHANGES_REQUESTED** [review]
Maybe we shall add how to extract all files from an update container, change them and re-create the update container?

**Comment by stulle123** [comment]
> Maybe we shall add how to extract all files from an update container, change them and re-create the update container?

Yes, that would be easier that dumping and reflashing the NAND. Just download a firmware upgrade from the vendor website, change it, "re-package" and flash via the web interface.

File: fw_upgrade/README.md (hunk 2 above)
Comment (stulle123) [reply fw_upgrade/README.md:37]: Fixed.

File: fw_upgrade/README.md (hunk 3 above)
Comment (stulle123) [reply fw_upgrade/README.md:54]: Fixed.

File: fw_upgrade/README.md (hunk 1 above)
Comment (stulle123) [reply fw_upgrade/README.md:16]: Fixed.

**Pull request merged**